File: src/main/java/ru/highloadcup/dao/UserDao.java
package ru.highloadcup.dao;

import org.jooq.DSLContext;
import org.jooq.Query;
import org.jooq.Record;
import org.jooq.RecordMapper;
import org.jooq.impl.DSL;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import ru.highloadcup.api.User;
import ru.highloadcup.generated.tables.records.UserRecord;

import javax.transaction.Transactional;


import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

import static ru.highloadcup.generated.Tables.USER;

@Component
public class UserDao {

    @Autowired
    private DSLContext dsl;

    @Transactional
    public int createUser(User user) {
        return dsl.insertInto(USER)
                .set(USER.ID, user.getId())
                .set(USER.EMAIL, user.getEmail())
                .set(USER.FIRST_NAME, user.getFirstName())
                .set(USER.LAST_NAME, user.getLastName())
                .set(USER.GENDER, user.getGender().name())
                .set(DSL.field("BIRTH_DATE", Long.class), user.getBirthDate())
                .execute();
    }

    @Transactional
    public void createUsers(Collection<User> users) {
        List<Query> queries = new ArrayList<>();
        for (User user : users) {
            queries.add(dsl.insertInto(USER)
                    .set(USER.ID, user.getId())
                    .set(USER.EMAIL, user.getEmail())
                    .set(USER.FIRST_NAME, user.getFirstName())
                    .set(USER.LAST_NAME, user.getLastName())
                    .set(USER.GENDER, user.getGender().name())
                    .set(DSL.field("BIRTH_DATE", Long.class), user.getBirthDate()));
        }
        dsl.batch(queries).execute();
    }

    @Transactional
    public int updateUser(Integer userId, User user) {
        UserRecord userRecord = dsl.fetchOne(USER, USER.ID.equal(userId));
        if (userRecord == null) {
            return 0;
        }
        if (user.getEmail() != null) {
            userRecord.setEmail(user.getEmail());
        }
        if (user.getFirstName() != null) {
            userRecord.setFirstName(user.getFirstName());
        }
        if (user.getLastName() != null) {
            userRecord.setLastName(user.getLastName());
        }
        if (user.getGender() != null) {
            userRecord.setGender(user.getGender().name());
        }
        if (user.getBirthDate() != null) {
            userRecord.set(DSL.field("BIRTH_DATE", Long.class), user.getBirthDate());
        }
        return userRecord.store();
    }

    public User getUser(Integer userId) {
        return dsl.select(USER.ID,
                USER.EMAIL,
                USER.FIRST_NAME,
                USER.LAST_NAME,
                USER.GENDER,
                USER.BIRTH_DATE)
                .from(USER)
                .where(USER.ID.equal(userId))
                .fetchOne(UserMapper.INSTANCE);
    }

    private static class UserMapper implements RecordMapper<Record, User> {
        private static final UserMapper INSTANCE = new UserMapper();

        @Override
        public User map(Record record) {
            User user = new User();
            user.setId(record.getValue(USER.ID));
            user.setEmail(record.getValue(USER.EMAIL));
            user.setFirstName(record.getValue(USER.FIRST_NAME));
            user.setLastName(record.getValue(USER.LAST_NAME));
            user.setGender(User.Gender.valueOf(record.getValue(USER.GENDER)));
            user.setBirthDate(record.getValue(USER.BIRTH_DATE, Long.class));
            return user;
        }
    }
}

File: src/main/java/ru/highloadcup/dao/UserDao.kt
package ru.highloadcup.dao

import org.jooq.DSLContext
import org.jooq.Query
import org.jooq.Record
import org.jooq.RecordMapper
import org.jooq.impl.DSL
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import ru.highloadcup.AllOpen
import ru.highloadcup.api.User
import ru.highloadcup.api.UserEO

import javax.transaction.Transactional


import java.util.ArrayList

import ru.highloadcup.generated.Tables.USER

@AllOpen
@Component
class UserDao {

    @Autowired
    private lateinit var dsl: DSLContext

    @Transactional
    fun createUser(user: User) = dsl.insertInto(USER)
            .set(USER.ID, user.id)
            .set(USER.EMAIL, user.email)
            .set(USER.FIRST_NAME, user.first_name)
            .set(USER.LAST_NAME, user.last_name)
            .set(USER.GENDER, user.gender.name)
            .set(DSL.field("BIRTH_DATE", Long::class.java), user.birthDate)
            .execute()

    @Transactional
    fun createUsers(users: Collection<User>) {
        val queries = users.mapTo(ArrayList<Query>()) { dsl.insertInto(USER)
                    .set(USER.ID, it.id)
                    .set(USER.EMAIL, it.email)
                    .set(USER.FIRST_NAME, it.first_name)
                    .set(USER.LAST_NAME, it.last_name)
                    .set(USER.GENDER, it.gender.name)
                    .set(DSL.field("BIRTH_DATE", Long::class.java), it.birthDate)
        }
        dsl.batch(queries).execute()
    }

    @Transactional
    fun updateUser(userId: Int, user: UserEO): Int {
        val userRecord = dsl.fetchOne(USER, USER.ID.equal(userId)) ?: return 0
        user.email?.let { userRecord.email = it }
        user.firstName?.let { userRecord.firstName = it }
        user.lastName?.let { userRecord.lastName = it }
        user.gender?.let { userRecord.gender = it.name }
        user.birthDate?.let { userRecord.set(DSL.field("BIRTH_DATE", Long::class.java), it) }
        return userRecord.store()
    }

    fun getUser(userId: Int) = dsl.select(USER.ID,
            USER.EMAIL,
            USER.FIRST_NAME,
            USER.LAST_NAME,
            USER.GENDER,
            USER.BIRTH_DATE)
            .from(USER)
            .where(USER.ID.equal(userId))
            .fetchOne(UserMapper)

    private object UserMapper : RecordMapper<Record, User> {
        override fun map(record: Record): User {
            return User(id = record.getValue(USER.ID),
                    email = record.getValue(USER.EMAIL),
                    first_name = record.getValue(USER.FIRST_NAME),
                    last_name = record.getValue(USER.LAST_NAME),
                    gender = User.Gender.valueOf(record.getValue(USER.GENDER)),
                    birthDate = record.getValue(USER.BIRTH_DATE, Long::class.java))
        }
    }
}
